source/simulator/helper_functions.cc

```diff
@@ -909,8 +909,164 @@ namespace aspect
+              {
+                min_advection_local = std::min<double> (min_advection_local, values_12[q]);
+                max_advection_local = std::max<double> (max_advection_local, values_12[q]);
+              }
+            for (unsigned int q = 0; q < n_q_points_21; ++q)
+              {
+                min_advection_local = std::min<double> (min_advection_local, values_21[q]);
+                max_advection_local = std::max<double> (max_advection_local, values_21[q]);
+              }
+
+            //Find the trouble cell
+            if (   min_advection_local < min_advection_exact_global
+                   || max_advection_local > max_advection_exact_global )
+              {
+                //Compute the cell area and cell advection average
+                double local_area = 0;
+                for (unsigned int q = 0; q < n_q_points_12; ++q)
+                  {
+                    local_area += fe_values_12.JxW(q);
+                  }
+
+
+                double local_advection_average = 0;
+                for (unsigned int q = 0; q < n_q_points_12; ++q)
+                  {
+                    local_advection_average += values_12[q]*fe_values_12.JxW(q);
+                  }
+                local_advection_average /= local_area;
+                //Define theta: which is a scaling constant
+                double theta_T = std::min<double>
+                                 (1, abs((max_advection_exact_global-local_advection_average)
+                                         /(max_advection_local-local_advection_average)));
+                theta_T = std::min<double>
+                          (theta_T, abs((min_advection_exact_global-local_advection_average)
+                                        /(min_advection_local-local_advection_average)));
+                //Modify the numerical solution at freedoms of advection field
+                for (unsigned int j = 0;
+                     j < finite_element.base_element(advection_field.base_element(introspection)).dofs_per_cell;
+                     ++j)
+                  {
+                    const unsigned int support_point_index = finite_element.component_to_system_index(
+                                                               (advection_field.is_temperature()
+                                                                ?
+                                                                introspection.component_indices.temperature
+                                                                :
+                                                                introspection.component_indices.compositional_fields[advection_field.compositional_variable]
+                                                               ),
+                                                               /*dof index within component=*/ j);
+                    double t_tmp = distributed_vector(local_dof_indices[support_point_index]);
```

Comment:
`t_tmp` is not a very helpful variable name. What about my suggestion to rename it to `solution_value` or something similar and name the variable in the next line `limited_solution_value` or `bounded_solution_value` and make both `const double`?